### PR TITLE
Validate base membership exists for lab access purchase

### LINKED ISSUE #179: Warn/block member from buying lab membership if person does not have yearly membership
# Current behaviour
Member can buy yearly membership and lab membership independently of each other.

# Wanted behaviour
Member should be warned or blocked from buying lab membership if they do not have active yearly membership

### PR DESCRIPTION
**Pull request by velaco**
Work in progress. Leaving placeholder for new tests to add

Fixes #179 

## Summary by CodeRabbit

- **New Features**
	- Enhanced cart processing logic to enforce membership requirements for lab access purchases.

- **Tests**
	- Added comprehensive test cases to validate lab access purchase scenarios, including:
		- Preventing lab access purchase without base membership
		- Allowing lab access purchase with base membership
		- Verifying lab access purchase with an active membership

### CODE REVIEW

**Comment by velaco** [comment]
Added some tests, but second one failing on payment part. Need to fix

File: api/src/shop/transactions.py [hunk 1]
```diff
@@ -342,11 +342,14 @@ def payment_success(transaction: Transaction) -> None:
...
 def process_cart(member_id: int, cart: List[CartItem]) -> Tuple[Decimal, List[TransactionContent]]:
     contents = []
+    labaccess_in_cart = False
+    base_membership_in_cart = False
 
     member = db_session.get(Member, member_id)
     if member is None:
         raise NotFound(message=f"Could not find member with id {member_id}.")
     price_level = get_price_level_for_member(member)
+    member_has_base_membership = Span.MEMBERSHIP in [span.type for span in member.spans]
```
Comment (emanuelen5) [inline api/src/shop/transactions.py:352]: Does this check really work if the base membership span has ended? Like a person buys base membership one year, then the next year when the base membership has ended, then, would they be able to buy labaccess?

File: api/src/shop/transactions.py [hunk 2]
```diff
@@ -395,6 +406,11 @@ def process_cart(member_id: int, cart: List[CartItem]) -> Tuple[Decimal, List[Tr
             content = TransactionContent(product_id=product_id, count=count, amount=amount)
             contents.append(content)
 
+        if labaccess_in_cart and not (member_has_base_membership or base_membership_in_cart):
+            raise BadRequest(
+                "Could not purchase selected subscription. Please buy the base annual membership before lab access."
+            )
+
```
Comment (emanuelen5) [inline api/src/shop/transactions.py:413]: I would prefer to break out this business rule as a function instead, which would make it possible to "conditionalize" easier and see what is "business logic".

**Review by emanuelen5 — APPROVED** [review]

**Comment by HalfVoxel** [comment]
I don't like the hackiness of this PR, and I think it should be reverted. I wasn't at the hack day yesterday, so didn't see this until now.

1. There is already functionality for rejecting purchases based on rules. They are in the `api/src/shop/filter.py` file. It would be good if all filtering is done using the same system. Possibly that system can be extended.
2. The hardcoded product IDs do not necessarily correspond with products giving labacccess/membership. The better way to check this is if the product has an action that gives labaccess or membership. But again, I think it should be done using the filtering system mention in (1).
3. The "member_has_base_membership" check is not correct, since a span could exist only in the future, but the member may not have access right now. Your check also does not handle deleted spans, which are not too unusual. It's better to use existing functions like `get_membership_summary(member_id)` which handle this properly.

**Pull request merged**